### DISCUSSION TITLE
Opendrain GPIO on STM32Fxx Microcontrollers

### DISCUSSION THREAD

**Discussion by findus4fun** [terjeio/grblHAL#175] [General]
Hello,

is there a way to set the GPIO outputs of the STM32F1/STM32F4 to OPEN DRAIN in grblHAL like mentioned in

https://www.st.com/resource/en/application_note/dm00315319-stm32-gpio-configuration-for-hardware-settings-and-lowpower-consumption-stmicroelectronics.pdf
section 3.2.2?

Background: I have an old breakoutboard with arduino NANO which has optos that are connected to +5V and are pulled low by arduino. I think I could leave the breakoutboard using 5V if it is possible to use open drain configuration of STM32Fxx GPIO pins.

Maybe somebody can give me a hint.

Regards
Jan

**Comment by terjeio** [comment]
> is there a way to set the GPIO outputs of the STM32F1/STM32F4 to OPEN DRAIN in grblHAL

Currently no - but should be fairly easy to implement as a driver specific #define compile time option by changing
`GPIO_Init.Mode = GPIO_MODE_OUTPUT_PP;`
to
`GPIO_Init.Mode = GPIO_MODE_OUTPUT_OD;`

However, a careful check of the datasheet for the MCU needs to be done first as the F411 has at least one pin that is not listed as FT (FT = five volt tolerant). I have not checked other variants.

**Reply by findus4fun** [reply]
Thanks for this information. I will have a closer look at the datasheets and search

GPIO_Init.Mode = GPIO_MODE_OUTPUT_PP;

in the code.

According to https://docs.zephyrproject.org/latest/_images/Blackpill_Pinout1.png

Pin A0 and B5 are not 5V tolerant on STM32F411

**Comment by findus4fun** [comment]
Hi,
what do you think if we added something like 

```
#define STEP_GPIO_OUT_MODE   GPIO_MODE_OUTPUT_OD
#define DIR_GPIO_OUT_MODE    GPIO_MODE_OUTPUT_OD
#define ENABLE_GPIO_OUT_MODE GPIO_MODE_OUTPUT_OD
```

in the map.h files and also add something like

```
#ifndef STEP_GPIO_OUT_MODE
#define STEP_GPIO_OUT_MODE GPIO_MODE_OUTPUT_PP
#endif 
```

in the driver.h files and replace 

`GPIO_Init.Mode = GPIO_MODE_OUTPUT_PP;`

with

`GPIO_Init.Mode = STEP_GPIO_OUT_MODE`

at the appropriate lines of the driver.c files.

That way it would be backwards compatible and leave everybody the freedom to use Push Pulll or Open Drain for the GPIO Out pins.

What do you think?

**Reply by terjeio** [reply]
Hmm - maybe better to add a #define which makes the intention clearer?

Add to driver maps that needs them, or perhaps better to `my_machine.h` as an option since as a general rule the map files should not normally be touched by users:

`#define ENABLE_5V_TOLERANT_OUTPUTS`

And in driver.c the few places this is needed:
```
#ifdef ENABLE_5V_TOLERANT_OUTPUTS
GPIO_Init.Mode = GPIO_MODE_OUTPUT_OD;
#else
GPIO_Init.Mode = GPIO_MODE_OUTPUT_PP;
#endif
```

What do you think?

**Comment by findus4fun** [comment]
Okay, 

I think your solution seems easier to implement. I will give it a try.

**Reply by terjeio** [reply]
Good, please make any PR against the test branch as the master is far behind.
